bd-resilient-kv/src/versioned_recovery.rs

```diff
@@ -0,0 +1,356 @@
+// shared-core - bitdrift's common client/server libraries
+// Copyright Bitdrift, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+use crate::kv_journal::TimestampedValue;
+use ahash::AHashMap;
+use bd_bonjson::Value;
+use bd_bonjson::decoder::from_slice;
+use flate2::read::ZlibDecoder;
+use std::io::Read;
+use std::path::Path;
+
+/// Helper function to read a u64 field from a BONJSON object.
+///
+/// BONJSON's decoder automatically converts unsigned values that fit in i64 to signed values
+/// during decoding (see bd-bonjson/src/decoder.rs:227-234). This means that even though we
+/// write `Value::Unsigned(version)`, the decoder returns `Value::Signed(version as i64)`.
+///
+/// TODO(snowp): Consider changing BONJSON's decoder to preserve the original unsigned type
+/// to avoid this normalization behavior and eliminate the need for this helper.
+fn read_u64_field(obj: &AHashMap<String, Value>, key: &str) -> Option<u64> {
+  match obj.get(key) {
+    Some(Value::Unsigned(v)) => Some(*v),
+    Some(Value::Signed(v)) if *v >= 0 =>
+    {
+      #[allow(clippy::cast_sign_loss)]
+      Some(*v as u64)
+    },
+    _ => None,
+  }
+}
+
+/// A utility for recovering state at arbitrary versions from raw journal data.
+///
+/// This utility operates on raw byte slices from versioned journals and can reconstruct
+/// the key-value state at any historical version by replaying journal entries.
+///
+/// Supports both compressed (zlib) and uncompressed journals. Compressed journals are
+/// automatically detected and decompressed transparently.
+///
+/// # Usage
+///
+/// ```ignore
+/// use bd_resilient_kv::VersionedRecovery;
+///
+/// // Load journal data as byte slices (may be compressed or uncompressed)
+/// let archived_journal = std::fs::read("store.jrn.v30000.zz")?; // Compressed
+/// let active_journal = std::fs::read("store.jrn")?; // Uncompressed
+///
+/// // Create recovery utility with both journals
+/// let recovery = VersionedRecovery::new(vec![&archived_journal, &active_journal])?;
+///
+/// // Recover state at specific version
+/// let state_at_25000 = recovery.recover_at_version(25000)?;
+/// ```
+#[derive(Debug)]
+pub struct VersionedRecovery {
+  journals: Vec<JournalInfo>,
+}
+
+#[derive(Debug)]
+struct JournalInfo {
+  data: Vec<u8>,
+  base_version: u64,
+  max_version: u64,
+}
+
+impl VersionedRecovery {
```

Comment:
Seems like this code was generated to do basically what I said about restart of reading all files, so wondering what the compaction is needed for.